### ISSUE TITLE
(Time-series) Regularization by lengthscale, likelihood noise or number of inducing points?

### ISSUE THREAD

**Issue by ginward** [cornellius-gp/gpytorch#1445]
Hello, 

I am currently working with time-series data with switching regimes. This means that GP might fit well in the training period (20% R2) but fit very poorly in the OOS period (-20% R2). 

I was wondering which regularization technique is most promising? I can see that both kernel lengthscale and likelihood noise levels could affect the fit level, and I can set constraints for them:

```
gpytorch.kernels.RBFKernel(lengthscale_constraint=gpytorch.constraints.GreaterThan(1))

likelihood = gpytorch.likelihoods.GaussianLikelihood(noise_constraint=gpytorch.constraints.GreaterThan(1))

```

But I am not sure how to set those constraints, because it doesn't seem like a GP thing to do cross-validation (from my limited understanding, you would recommend using `mll` on the training sample in training model, which replaces cross-validation. However, how am I suppose to select the hyperparameter in this case? 

Also, if I am doing approximate GP, will the number of inducing point potentially affect overfitting, too? 

Thanks!

**Comment by jacobrgardner** [comment]
With time series data, is your test set actually like extrapolation (or even a large contiguous region where you are interpolating) or something? In those cases, simple RBF kernels are just going to do really poorly and you need to use specific modelling tools to do time series analysis with GPs.

See my response to #1444 for early stopping.

**Comment by ginward** [comment]
> With time series data, is your test set actually like extrapolation (or even a large contiguous region where you are interpolating) or something? In those cases, simple RBF kernels are just going to do really poorly and you need to use specific modelling tools to do time series analysis with GPs.
> 
> See my response to #1444 for early stopping.

Thanks for your quick response! 

It is a panel data of thousands of stocks, and I am using firm characteristics to predict stock returns. I think it is sufficient to say that this process is not stationary. 

I am currently using a rolling window + GP, with about 1 year of training and 3 months of testing data. 

What type of kernel do you recoomend in this case?

**Comment by ginward** [comment]
By the way, is there way to do L2 regularisation (for example, penalise the kernel matrix) in GPytorch? Does the noise parameter do the job?

**Comment by ginward** [comment]
> By the way, is there way to do L2 regularisation (for example, penalise the kernel matrix) in GPytorch? Does the noise or lengthscale parameter do the job?

For example, does adding weight_decay to the Adam optimiser work well? 

```
    optimizer = t.optim.Adam([
      #{'params': model.feature_extractor.parameters()},
      {'params': model.covar_module.parameters()},
      {'params': model.mean_module.parameters()},
      {'params': model.likelihood.parameters()},
    ], lr=0.1, weight_decay = 1e-1)
```

**Comment by gpleiss** [comment]
> What type of kernel do you recoomend in this case?

A spectral mixture kernel should work well.

> By the way, is there way to do L2 regularisation (for example, penalise the kernel matrix) in GPytorch? Does the noise or lengthscale parameter do the job?

For bayesian modeling, it doesn't really make sense to have L2 regularization. Complexity in the kernel matrix is penalized by the log determinant term in the marginal log likelihood (see Rasmussen & Williams, Ch 5).

In the future - this would make for a better [Discussion topic](https://github.com/cornellius-gp/gpytorch/discussions) rather than an issue. I'm closing this for now, and please reopen a discussion if you wish to talk about this further.
